packages/react-reconciler/src/ReactStrictModeWarnings.js

```diff
@@ -289,6 +296,70 @@ if (__DEV__) {
       });
     }
   };
+
+  ReactStrictModeWarnings.recordLegacyContextWarning = (
+    fiber: Fiber,
+    instance: any,
+  ) => {
+    // Dedup strategy: Warn once per component.
+    if (didWarnAboutLegacyContext.has(fiber.type)) {
+      return;
+    }
+
+    const strictRoot = findStrictRoot(fiber);
+    if (strictRoot === null) {
+      warning(
+        false,
+        'Expected to find a StrictMode component in a strict mode tree. ' +
+          'This error is likely caused by a bug in React. Please file an issue.',
+      );
+      return;
+    }
+
+    let warningsForRoot = pendingLegacyContextWarning.get(strictRoot);
+    if (warningsForRoot === undefined) {
+      warningsForRoot = [];
+      pendingLegacyContextWarning.set(strictRoot, warningsForRoot);
+    }
```

Comment:
Can we not initialize the `Set` until we actually have something to warn about? (In other words, can we move this initialization into the conditional below?) Seems like that would be a little better for perf.

I think this would also let us skip the  `fiberArray.length > 0` check in the `forEach` below.

Reply:
That's a good point.